[jit] dropout symbolic_script should respect the training flag

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20760 [jit] dropout symbolic_script should respect the training flag**

as title

Differential Revision: [D15486511](https://our.internmc.facebook.com/intern/diff/D15486511)

suo committed:
[jit] dropout symbolic_script should respect the training flag

as title

suo committed:
Update on "[jit] dropout symbolic_script should respect the training flag"

[jit] dropout symbolic_script should respect the training flag

as title

gh-metadata: pytorch pytorch 20760 gh/suo/41/head

suo committed:
Update on "[jit] dropout symbolic_script should respect the training flag"

[jit] dropout symbolic_script should respect the training flag

as title

gh-metadata: pytorch pytorch 20760 gh/suo/41/head

ailzhang commented:
is it possible to add a test to prevent regression in the future?

zdevito commented on torch/csrc/jit/symbolic_script.cpp:
```diff
@@ -1116,18 +1116,29 @@ const std::vector<std::string> functions = {
-                res = mask * input / p1m
+                p1m = 1.
+                res = input
+                mask = torch.zeros(0)
```
The definition of backwards when is_training = False is not correct.

suo replied:
oops this got merged out somehow

suo committed:
Update on "[jit] dropout symbolic_script should respect the training flag"

[jit] dropout symbolic_script should respect the training flag

as title

gh-metadata: pytorch pytorch 20760 gh/suo/41/head

suo committed:
Update on "[jit] dropout symbolic_script should respect the training flag"

[jit] dropout symbolic_script should respect the training flag

as title. This unfortunately means that the forward for dropout doesn't
fuse completely anymore, but the "important" parts are fused, and all
we're adding is the is_training check overhead. The only time we're
doing "extra" stuff is if we 1) require_grad and 2) are not training,
which seems like uncommon things.

gh-metadata: pytorch pytorch 20760 gh/suo/41/head

suo committed:
Update on "[jit] dropout symbolic_script should respect the training flag"

[jit] dropout symbolic_script should respect the training flag

as title. This unfortunately means that the forward for dropout doesn't
fuse completely anymore, but the "important" parts are fused, and all
we're adding is the is_training check overhead. The only time we're
doing "extra" stuff is if we 1) require_grad and 2) are not training,
which seems like uncommon things.

gh-metadata: pytorch pytorch 20760 gh/suo/41/head
